# Import libraries


In [32]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import seaborn as sns
import mysql.connector
import sqlalchemy

## Load data

The dataset has three tables, containing offer data, customer profile and transaction data. They will be assigned to DataFrames df_offer, df_customer and df_transcript.

In [2]:
transcript = pd.read_csv('transcript.csv')
profile = pd.read_csv('profile.csv')
pf = pd.read_csv('portfolio.csv')

In [14]:
df_offer.head()

reward                              channels  difficulty  duration  \
0      10         ['email', 'mobile', 'social']          10         7   
1      10  ['web', 'email', 'mobile', 'social']          10         5   
2       0            ['web', 'email', 'mobile']           0         4   
3       5            ['web', 'email', 'mobile']           5         7   
4       5                      ['web', 'email']          20        10   

      offer_type                                id  
0           bogo  ae264e3637204a6fb9bb56bc8210ddfd  
1           bogo  4d5c57ea9a6940dd891ad53e9dbe8da0  
2  informational  3f207df678b143eea3cee63160fa8bed  
3           bogo  9b98b8c7a33c4b65b9aebfe6a799e6d9  
4       discount  0b1e1539f2cc45b7b9fa7c272da2e1d7

In [15]:
df_customer.head()

gender  age                                id  became_member_on    income
0    NaN  118  68be06ca386d4c31939f3a4f0e3dd783          20170212       NaN
1      F   55  0610b486422d4921ae7d2bf64640c50b          20170715  112000.0
2    NaN  118  38fe809add3b4fcf9315a9694bb96ff5          20180712       NaN
3      F   75  78afa995795e4d85b5d9ceeca43f5fef          20170509  100000.0
4    NaN  118  a03223e636434f42ac4c3df47e8bac43          20170804       NaN

In [16]:
df_transcript.head()

person           event  \
0  78afa995795e4d85b5d9ceeca43f5fef  offer received   
1  a03223e636434f42ac4c3df47e8bac43  offer received   
2  e2127556f4f64592b11af22de27a7932  offer received   
3  8ec6ce2a7e7949b1bf142def7d0e0586  offer received   
4  68617ca6246f4fbc85e91a2a49552598  offer received   

                                              value  time  
0  {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}     0  
1  {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}     0  
2  {'offer id': '2906b810c7d4411798c6938adc9daaa5'}     0  
3  {'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'}     0  
4  {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}     0

## Data Cleaning

### Drop the unnecessary column(s)

Column 'Unnamed: 0' is not useful hence to be dropped.

In [3]:
df_offer = pf.drop('Unnamed: 0', axis = 1)
df_customer = profile.drop('Unnamed: 0', axis = 1)
df_transcript = transcript.drop('Unnamed: 0', axis = 1)

### Check for missing values

In [4]:
df_offer.isna().sum()

reward        0
channels      0
difficulty    0
duration      0
offer_type    0
id            0
dtype: int64

In [5]:
df_customer.isna().sum()

gender              2175
age                    0
id                     0
became_member_on       0
income              2175
dtype: int64

In [6]:
df_transcript.isna().sum()

person    0
event     0
value     0
time      0
dtype: int64

There is no missing value in df_offer and df_transcript.

For df_customer, there are 2175 missing values in column gender and the same number of missing values in column income. Need to check whether these missing values are from the same observations.

In [7]:
missing_gender = df_customer[df_customer['gender'].isna()]
missing_income = df_customer[df_customer['income'].isna()]

np.sum(missing_gender['id'] == missing_income['id'])

2175

All missing values in columns gender and income are from the same observations.

income is likely to play an important role in customer behaviour, hence observations with missing income will be removed from the dataset. This means 12.8% of the observations will be removed which is not ideal, however the alternative (replacing NaNs with the average or the median) will make our analysis less robust. As there are enough obsevations in our data for analysis, this should not have an important impact on the analysis.

In [11]:
id_to_remove = missing_income['id']
df_customer_no_na = df_customer[~df_customer['id'].isin(id_to_remove)]
df_customer_no_na = df_customer_no_na.reset_index(drop = True)
df_customer_no_na.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14825 entries, 0 to 14824
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            14825 non-null  object 
 1   age               14825 non-null  int64  
 2   id                14825 non-null  object 
 3   became_member_on  14825 non-null  int64  
 4   income            14825 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 579.2+ KB


There are no missing values in df_customer_no_na.

## Check data types

In [18]:
df_offer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   reward      10 non-null     int64 
 1   channels    10 non-null     object
 2   difficulty  10 non-null     int64 
 3   duration    10 non-null     int64 
 4   offer_type  10 non-null     object
 5   id          10 non-null     object
dtypes: int64(3), object(3)
memory usage: 608.0+ bytes


In [19]:
df_customer_no_na.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14825 entries, 0 to 14824
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            14825 non-null  object 
 1   age               14825 non-null  int64  
 2   id                14825 non-null  object 
 3   became_member_on  14825 non-null  int64  
 4   income            14825 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 579.2+ KB


Column became_member_on are integers, which should be coverted to datetime type.

In [20]:
df_customer_no_na['became_member_on'] = pd.to_datetime(df_customer_no_na['became_member_on'], format = '%Y%m%d')

In [21]:
df_transcript.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306534 entries, 0 to 306533
Data columns (total 4 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   person  306534 non-null  object
 1   event   306534 non-null  object
 2   value   306534 non-null  object
 3   time    306534 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 9.4+ MB


In [22]:
# Rename column `time`
df_transcript.rename(columns = {'time' : 'hours_since_start'}, inplace = True)
df_transcript.head(1)

person           event  \
0  78afa995795e4d85b5d9ceeca43f5fef  offer received   

                                              value  hours_since_start  
0  {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}                  0

## Data Transformation

Add an alias to each offer for easier recognition and referencing.

In [23]:
df_offer = df_offer.sort_values(['offer_type', 'difficulty']).reset_index(drop = True)
from string import ascii_uppercase
df_offer['offer_alias'] = [ascii_uppercase[i] for i in range(df_offer.shape[0])]
df_offer

reward                              channels  difficulty  duration  \
0       5            ['web', 'email', 'mobile']           5         7   
1       5  ['web', 'email', 'mobile', 'social']           5         5   
2      10         ['email', 'mobile', 'social']          10         7   
3      10  ['web', 'email', 'mobile', 'social']          10         5   
4       3  ['web', 'email', 'mobile', 'social']           7         7   
5       2  ['web', 'email', 'mobile', 'social']          10        10   
6       2            ['web', 'email', 'mobile']          10         7   
7       5                      ['web', 'email']          20        10   
8       0            ['web', 'email', 'mobile']           0         4   
9       0         ['email', 'mobile', 'social']           0         3   

      offer_type                                id offer_alias  
0           bogo  9b98b8c7a33c4b65b9aebfe6a799e6d9           A  
1           bogo  f19421c1d4aa40978ebb69ca19b0e20d           B  
2           bogo  ae264e3637204a6fb9bb56bc8210ddfd           C  
3           bogo  4d5c57ea9a6940dd891ad53e9dbe8da0           D  
4       discount  2298d6c36e964ae4a3e7e9706d1fb8c2           E  
5       discount  fafdcd668e3743c1bb461111dcafc2a4           F  
6       discount  2906b810c7d4411798c6938adc9daaa5           G  
7       discount  0b1e1539f2cc45b7b9fa7c272da2e1d7           H  
8  informational  3f207df678b143eea3cee63160fa8bed           I  
9  informational  5a8bc65990b245e5a138643cd4eb9837           J

Elements of column value are dictionaries as strings. I will separate the keys and the values and place them into two columns for easier manipulation.

In [24]:
value_column = df_transcript['value']
value_column = value_column.apply(eval)
value_column[:3]

0    {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}
1    {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}
2    {'offer id': '2906b810c7d4411798c6938adc9daaa5'}
Name: value, dtype: object

In [25]:
dictionary_key_column = [list(d.keys())[0] for d in value_column]
dictionary_value_column = [list(d.values())[0] for d in value_column]

value_column_split = pd.DataFrame(columns = ['dict_key', 'dict_value'])
value_column_split['dict_key'] = dictionary_key_column
value_column_split['dict_value'] = dictionary_value_column
value_column_split.head(3)

dict_key                        dict_value
0  offer id  9b98b8c7a33c4b65b9aebfe6a799e6d9
1  offer id  0b1e1539f2cc45b7b9fa7c272da2e1d7
2  offer id  2906b810c7d4411798c6938adc9daaa5

In [26]:
df_transcript_value_mod = df_transcript.drop('value', axis = 1)
df_transcript_value_mod = pd.concat([df_transcript_value_mod, value_column_split], axis = 1)

In [27]:
df_transcript_value_mod.columns.values

array(['person', 'event', 'hours_since_start', 'dict_key', 'dict_value'],
      dtype=object)

In [28]:
col_names = ['person', 'event', 'dict_key', 'dict_value', 'hours_since_start']
df_transcript_value_mod = df_transcript_value_mod[col_names]
df_transcript_value_mod.sample(3)

person            event  dict_key  \
35479  1d832bb5ffc6420ea7cea524ccdd45e1  offer completed  offer_id   
95824  f3b9b9a4adfb432dbb1d572aff469e99      transaction    amount   
39460  d930673b135b4ef38943a9acc93e5f81      transaction    amount   

                             dict_value  hours_since_start  
35479  fafdcd668e3743c1bb461111dcafc2a4                 66  
95824                              9.33                252  
39460                             18.64                 84

Create a new df focusing on the time customers became members

In [30]:
df_member = df_customer_no_na[['id', 'became_member_on']]

# Add columns `year`, `month`, `day_of_week`
df_member = df_member.assign(year = df_member.became_member_on.dt.year,
                             month = df_member.became_member_on.dt.month,
                             day = df_member.became_member_on.dt.day,
                             month_name = df_member.became_member_on.dt.month_name(),
                             day_of_week = df_member.became_member_on.dt.weekday,
                             day_name = df_member.became_member_on.dt.day_name())
df_member.head(3)

id became_member_on  year  month  day  \
0  0610b486422d4921ae7d2bf64640c50b       2017-07-15  2017      7   15   
1  78afa995795e4d85b5d9ceeca43f5fef       2017-05-09  2017      5    9   
2  e2127556f4f64592b11af22de27a7932       2018-04-26  2018      4   26   

  month_name  day_of_week  day_name  
0       July            5  Saturday  
1        May            1   Tuesday  
2      April            3  Thursday

In [31]:
offers = df_offer.copy()
customers = df_customer_no_na.copy()
transcripts = df_transcript_value_mod.copy()
members = df_member.copy()

After data cleaning and creating new data frame, we will make a connection to MSQL database and we will save our cleaned data in database.

In [33]:
database_username = 'root'
database_password = 'ozge.12'
database_ip       = '127.0.0.1'
database_name     = 'starbucks_db'
database_connection = sqlalchemy.create_engine('mysql+mysqlconnector://{0}:{1}@{2}/{3}'.
                                               format(database_username, database_password, 
                                                      database_ip, database_name), pool_recycle=1, pool_timeout=57600).connect()

offers.to_sql(con=database_connection, name='offers', if_exists='append',chunksize=100)
customers.to_sql(con=database_connection, name='customers', if_exists='append',chunksize=100)
transcripts.to_sql(con=database_connection, name='transcripts', if_exists='append',chunksize=100)
members.to_sql(con=database_connection, name='members', if_exists='append',chunksize=100)
database_connection.close()